### PR TITLE
Bump golang version in go.mod to go 1.21

### LINKED ISSUE #792: Golang 1.21 support for vsphere CPI
Bump golang 1.21 for source code and test

### PR DESCRIPTION
**Pull request by DanielXiao**
Apply go mod tidy in the following folders:
1. root folder.
2. hack/tools.
3. test/e2e.

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #792

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: test/e2e/go.mod [hunk 1]
```diff
@@ -1,6 +1,8 @@
 module k8s.io/cloud-provider-vsphere/test/e2e
 
-go 1.20
+go 1.21
+
+toolchain go1.21.5
```
Comment (wyike) [inline test/e2e/go.mod:5]: This line is automatically generated if a lower version in go.mod than the expected toolchain, when running `go mod tidy` with golang 1.21 in the local.

Hi Yifeng, can you manually change line 3 from `go 1.20 `to `go 1.21` first and then run "go mod tidy"?  In this way we won't cause` toolchain go1.21.5` auto generated.

Comment (DanielXiao) [reply test/e2e/go.mod]: Thanks you. Done.

**Pull request merged**